core/directory.go

```diff
@@ -523,6 +523,52 @@ func (dir *Directory) WithPatch(ctx context.Context, patch string) (*Directory,
 	return dir, nil
 }
 
+func (dir *Directory) Search(ctx context.Context, opts SearchOpts, paths []string, globs []string) ([]*SearchResult, error) {
+	ref, err := getRefOrEvaluate(ctx, dir)
+	if err != nil {
+		return nil, err
+	}
+	if ref == nil {
+		// empty directory, i.e. llb.Scratch()
+		return []*SearchResult{}, nil
+	}
+
+	opt, ok := buildkit.CurrentOpOpts(ctx)
+	if !ok {
+		return nil, fmt.Errorf("no buildkit opts in context")
+	}
+	ctx = trace.ContextWithSpanContext(ctx, opt.CauseCtx)
+
+	bkSessionGroup, ok := buildkit.CurrentBuildkitSessionGroup(ctx)
+	if !ok {
+		return nil, fmt.Errorf("no buildkit session group in context")
+	}
+
+	results := []*SearchResult{}
+	err = MountRef(ctx, ref, bkSessionGroup, func(root string) error {
+		resolvedDir, err := containerdfs.RootPath(root, dir.Dir)
+		if err != nil {
+			return err
+		}
+		rgArgs := opts.RipgrepArgs()
+		for _, glob := range globs {
+			rgArgs = append(rgArgs, "--glob="+glob)
+		}
+		if len(paths) > 0 {
+			rgArgs = append(rgArgs, "--")
+			rgArgs = append(rgArgs, paths...)
```

Comment:
Paths could be set to absolute paths, right? In which case we'd escape out of the `resolvedDir`?

Also, while I saw in the docs that `rg` won't follow symlinks when recursing, it didn't explicitly say what it does if a path itself is a symlink. We definitely don't want to follow those in that case either

Reply:
Ok this was kind of a pain in the butt to fix, but here it is: https://github.com/dagger/dagger/pull/10937/commits/dd661fb163c7403299603d17f62b04ef67190bed - not sure if there's a cleaner approach. :thinking: